data/data_utils.py

```diff
@@ -0,0 +1,601 @@
+
+    label = curr_subgoal
+
+    input_str = task_desc + f" </s> Time: {step_id}; Score: {int(recent_scores[-1] * 100)}; </s> "
+
+    if len(subgoals) > 1:
+        input_str += '. Completed subtasks are: ' + ', '.join(sg for sg in subgoals[:-1])
+    else:
+        ### the completed subgoals are empty
+        input_str += '. No completed subtasks'
+
+    if len(subgoals) > 0:
+        input_str += '. The current subtask is ' + subgoals[-1]
+
+    input_str += "Action history: </s>"
+    ind = 10
+    for obs, action, reward in zip(recent_obs[-10:], recent_actions[-10:], recent_reward[-10:]):
+        input_str += f" <extra_id_{ind}> {formalize_action(action)} (+{int(reward * 100)}) --> {obs} | "
+        ind -= 1
+    input_str += " </s> "
+
+    input_str += "Current environment: " + look + " </s> "
+    input_str += "Current inventory: " + inventory + " </s> "
+
+    if places:
+        input_str += "Visited rooms: " + ", ".join(places) + ' </s> '
+
+    input_str += ' . What subtask should you do next? </s> '
+    input_str = sanitizeStr(input_str)
+    input_str = input_str.replace("(that is open)", "")
+    input_str = input_str.replace("(containing nothing)", "")
+
+    return input_str, label
+
+
+def compose_instance_v4_sgaslabel_v4(step_id, task_desc, returns_to_go, curr_action,
+                           curr_obs, inventory, look, prev_action, prev_obs,
+                           objects, places, recent_actions, recent_obs, recent_scores, recent_reward, subgoals, curr_subgoal):
+
+    ''' In this function the labels/predictions are the subgoals instead of the actions and it keeps the history of (action, observation)..'''
+
+    label = curr_subgoal
+
+    input_str = task_desc + f" </s> Time: {step_id}; Score: {int(recent_scores[-1] * 100)}; </s> "
+
+    if len(subgoals) > 1:
+        input_str += '. Completed subtasks are: ' + ', '.join(sg for sg in subgoals[:-2])
+    else:
+        ### the completed subgoals are empty
+        input_str += '. No completed subtasks'
+
+    input_str += '. The current subtask is ' + subgoals[-2]
+
+    input_str += "Action history: </s>"
+    ind = 10
+    for obs, action, reward in zip(recent_obs[-10:], recent_actions[-10:], recent_reward[-10:]):
+        input_str += f" <extra_id_{ind}> {formalize_action(action)} (+{int(reward * 100)}) --> {obs} | "
+        ind -= 1
+    input_str += " </s> "
+
+    input_str += "Current environment: " + look + " </s> "
+    input_str += "Current inventory: " + inventory + " </s> "
+
+    if places:
+        input_str += "Visited rooms: " + ", ".join(places) + ' </s> '
+
+    input_str += ' . What action should you do next? </s> '
+    input_str = sanitizeStr(input_str)
+    input_str = input_str.replace("(that is open)", "")
+    input_str = input_str.replace("(containing nothing)", "")
+
+    return input_str, label
+
+
+def compose_instance_v4_sg_nohist(step_id, task_desc, returns_to_go, curr_action,
```

Comment:
Do you think it would be easier to read if we combine these cases into one base class, and inherit it for specific changes? Looks like apart from the small change (eg history or no history here) there might be redundancy in the code

Reply:
No, because it would be harder to follow. Also, this gives more flexibility if any user wants to add a new format.